engineering/architecture/numbers.md

```diff
@@ -0,0 +1,106 @@
+---
+title: Number display and entry
+description: Goals and strategies for the display and entry of numerical data
+published: true
+date: 2022-03-15T00:00:00.000Z
+tags: 
+editor: markdown
+dateCreated: 2022-03-15T00:00:00.000Z
+---
+
+## Product goals
+
+- **(Goal A)** Mathesar is capable of displaying and accepting entry for all [numbers which Postgres is capable of storing](https://www.postgresql.org/docs/current/datatype-numeric.html).
+
+- **(Goal B)** Aside from the numbers, `NaN`, `-Infinity`, and `Infinity`, users are able to type all numbers into Mathesar exactly as Mathesar displays those numbers to them.
+
+- **(Goal C)** When a user enters data which cannot be accurately stored in the column exactly as-entered, then Mathesar stores the nearest acceptable value and clearly communicates to stored value to the user.
+
+    For example, a `float8` column cannot accurately store the value `9999999999999999` (because it will instead be stored as `1e+16`). In this example, Mathesar will store `1e+16`, displaying that value back to the user after entry.
+
+- **(Goal D)** Mathesar displays numbers and accepts entry of numbers in the user's desired locale, including the correct decimal separator and grouping separators.
+
+- **(Goal E)** Mathesar auto-formats numbers as the user types.
+
+- **(Goal F)** Mathesar accepts entry of numbers in scientific notation and displays numbers in scientific notation when the user is likely to prefer it. For example `1E100` instead of 1 followed by 100 zeros.
```

Comment:
This is engineering notation, not scientific notation ([quick link explaining the difference in case anyone else is interested](https://www.medcalc.org/manual/scientific-notation.php)).

Reply:
Is it? My understanding is that `1E100` is scientific notation whereas the same number would be represented with `10E99` in engineering notation (to make the exponent a multiple of three). Though realistically, nobody would want to use engineering notation for a number that large because the [SI prefixes](https://en.wikipedia.org/wiki/Metric_prefix#List_of_SI_prefixes) (the main reason to use engineering notation, in my experience) stop well before 99.

Reply:
Another way to frame my understanding... I'd say that engineering notation is a form of scientific notation, but not necessarily the other way around.

Reply:
Also, here is how the `Intl` API behaves:

```txt
> Intl.NumberFormat('en-US', { notation: 'scientific' }).format(1E100)
"1E100"
> Intl.NumberFormat('en-US', { notation: 'engineering' }).format(1E100)
"10E99"
>
```

Reply:
Never mind, I got `E` notation and Engineering notation mixed up (because Engineering starts with an E).

i do think it would be cool to eventually support rendering non-E-scientific notation, e.g. 35.2 × 10<sup>-7</sup>. Probably not anytime soon, though, it sounds complicated.

Reply:
Yeah we could eventually have a display option for rich formatting. Though (to be pedantic), I don't think we'd ever format the number in your example using an exponent of `-7` -- instead it would be be 3.52 × 10<sup>-6</sup> (in both scientific and engineering notation).

We may even want a display option to use engineering notation at some point.

For `notation`, we could have:

- `'auto'` (default)
- `'standard'`
- `'scientific'`
- `'scientific-rich'`
- `'engineering-rich'`

With the code I have so far, this would all be easy to support.

Reply:
Interestingly, for some locales `Intl` actually behave more akin to the format you were expecting. For example:

```ts
Intl.NumberFormat('sv', { notation: 'scientific' }).format(1E100) // --> "1×10^100"
```

That's why I wrote in the specs that the exponent separator is to always be "E", regardless of locale.

Reply:
> Though (to be pedantic), I don't think we'd ever format the number in your example using an exponent of `-7`

I didn't even notice that (I stole the example from the MedCalc docs I linked above), I agree the exponent should be `-6`.

I'm glad the current implementation will be extensible for all sorts of formatting, I like your suggestion notation options.